bigquery/arrow.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bigquery
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"math/big"
+
+	"cloud.google.com/go/civil"
+	"github.com/apache/arrow/go/v10/arrow"
+	"github.com/apache/arrow/go/v10/arrow/array"
+	"github.com/apache/arrow/go/v10/arrow/ipc"
+	"github.com/apache/arrow/go/v10/arrow/memory"
+)
+
+type arrowDecoder struct {
+	mem            *memory.GoAllocator
+	tableSchema    Schema
+	rawArrowSchema []byte
+	arrowSchema    *arrow.Schema
+}
+
+func newArrowDecoderFromSession(session *ReadSession, schema Schema) (*arrowDecoder, error) {
+	arrowSerializedSchema := session.SerializedArrowSchema
+	mem := memory.NewGoAllocator()
+	buf := bytes.NewBuffer(arrowSerializedSchema)
+	r, err := ipc.NewReader(buf, ipc.WithAllocator(mem))
+	if err != nil {
+		return nil, err
+	}
+
+	p := &arrowDecoder{
+		mem:            mem,
+		tableSchema:    schema,
+		rawArrowSchema: arrowSerializedSchema,
+		arrowSchema:    r.Schema(),
+	}
+	return p, nil
+}
+
+// decodeArrowRecords decodes BQ ArrowRecordBatch into a list of arrow.Record.
+func (ap *arrowDecoder) decodeArrowRecords(serializedArrowRecordBatch []byte) ([]arrow.Record, error) {
+	buf := bytes.NewBuffer(ap.rawArrowSchema)
+	buf.Write(serializedArrowRecordBatch)
+	r, err := ipc.NewReader(buf, ipc.WithAllocator(ap.mem), ipc.WithSchema(ap.arrowSchema))
+	if err != nil {
+		return nil, err
+	}
+	records := []arrow.Record{}
```

Comment:
If we're not exposing the arrow iterator to users, does it make sense to convert to bigquery values in the decoder here? It looks like you manage the Retain / Release lifecycle of the arrow resources in a variety of places.

Does immediate conversion obviate that complexity?

Reply:
Initially when the `ArrowIterator` and the `RowIterator` were one thing, the `arrow.Record` to `[][]bigquery.Value` conversion was happening on the same struct, but as this was divided, it was required to call Retain/Release ( which right now each one is only called once ).

Changing. this back will change some concepts, as the ArrowIterator maintains a `chan arrow.Record` and it's `Next` method blocks/waits for new items to arrive on that channel, while the `RowIterator` works by dequeueing a row from the `rows`  attribute. Having that separating makes the changes required on the `RowIterator` less invasive, at the cost of having to keep an eye on those memory reference counts.